### PR TITLE
feat: remove docker image when component deleted

### PR DESCRIPTION
**Pull request by indougnito**
Copied from PR #1352:

**Description of changes:**
When a component which use docker image has been removed, check whether this image being used by other running components，if not, delete this image with docker cli commend rmi.

**Why is this change necessary:**
Before this change, unused images would be left and take lots of space.

**How was this change tested:**
- [x] Updated or added new unit tests.
- [x] Manual test.

**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/test/resources/com/aws/greengrass/componentmanager/recipes/Component_recipe_content.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  aws.greengrass.TokenExchangeService:
+    VersionRequirement: '>=2.0.0 <2.3.0'
+    DependencyType: SOFT
+Manifests:
+  - Platform:
+      os: all
+    Lifecycle:
+      Run: >-
+        docker run
+        450817829141.dkr.ecr.us-east-1.amazonaws.com/integrationdockerimage:latest
```
Comment (MikeDombo) [inline src/test/resources/com/aws/greengrass/componentmanager/recipes/Component_recipe_content.yaml:20]: do we really need to try running? This won't work on windows

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java [hunk 2]
```diff
@@ -279,4 +289,54 @@ private <T> T runWithConnectionErrorCheck(CrashableSupplier<T, Exception> task)
...
+    public boolean ifImageUsedByOther(ComponentStore componentStore) throws PackageLoadingException {
+        Map<String, Set<String>> allVersions = componentStore.listAvailableComponentVersions();
+        for (Map.Entry<String, Set<String>> versions : allVersions.entrySet()) {
+            String compName = versions.getKey();
+            Set<String> localVersions = new HashSet<>(versions.getValue());
+            for (String compVersion : localVersions) {
+                try {
+                    ComponentIdentifier identifier = new ComponentIdentifier(compName, new Semver(compVersion));
+                    // this.identifier is to be deleted identifier
+                    if (identifier.compareTo(this.identifier) == 0) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java:327]: just use `equals`? instead of compareto

**Review by junfuchen99 — APPROVED** [review]

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java [hunk 3]
```diff
@@ -226,15 +230,42 @@ ComponentRecipe getPackageRecipe(@NonNull ComponentIdentifier pkgId) throws Pack
      * @param compId component identifier
      * @throws PackageLoadingException if deletion of the component failed
      */
-    void deleteComponent(@NonNull ComponentIdentifier compId) throws PackageLoadingException {
+    void deleteComponent(@NonNull ComponentIdentifier compId,
+                         @NonNull ArtifactDownloaderFactory artifactDownloaderFactory)
+            throws PackageLoadingException, InvalidArtifactUriException {
         logger.atDebug("delete-component-start").kv("componentIdentifier", compId).log();
         IOException exception = null;
+        // issues #1111
```
Comment (Xuqing888) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:238]: Do we want to remove this comment?

**Review by Xuqing888 — APPROVED** [review]

**Review by MikeDombo — APPROVED** [review]

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java (hunk 3 above)
Comment (indougnito) [reply src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:238]: No strong opinion from me. Maybe it doesn't hurt to keep a ref to the issue number?

**Pull request merged**